src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeWalker.java

```diff
@@ -0,0 +1,305 @@
+        final int[] leftNodePositions;
+        // holds the address, in the packed byte[] index, of the right-node of each level:
+        final int[] rightNodePositions;
+        // holds the packed per-level split values; the run method uses this to save the cell min/max as it recurses:
+        final byte[][] splitValuesStack;
+
+        int nodeID;
+        int level;
+
+        TraversalState(DataInput dataInput) throws IOException
+        {
+            maxPointsInLeafNode = dataInput.readVInt();
+            bytesPerValue = dataInput.readVInt();
+
+            // Read index:
+            numLeaves = dataInput.readVInt();
+            assert numLeaves > 0;
+            treeDepth = dataInput.readVInt();
+            minPackedValue = new byte[bytesPerValue];
+            maxPackedValue = new byte[bytesPerValue];
+
+            dataInput.readBytes(minPackedValue, 0, bytesPerValue);
+            dataInput.readBytes(maxPackedValue, 0, bytesPerValue);
+
+            if (FutureArrays.compareUnsigned(minPackedValue, 0, bytesPerValue, maxPackedValue, 0, bytesPerValue) > 0)
+            {
+                String message = String.format("Min packed value %s is > max packed value %s.",
+                                               new BytesRef(minPackedValue), new BytesRef(maxPackedValue));
+                throw new CorruptIndexException(message, dataInput);
+            }
+
+            pointCount = dataInput.readVLong();
+
+            int numBytes = dataInput.readVInt();
+            byte[] packedIndex = new byte[numBytes];
+            dataInput.readBytes(packedIndex, 0, numBytes);
+
+            memoryUsage = ObjectSizes.sizeOfArray(packedIndex) + ObjectSizes.sizeOfArray(minPackedValue) + ObjectSizes.sizeOfArray(maxPackedValue);
+
+            nodeID = 1;
+            level = 0;
+            leafBlockFPStack = new long[treeDepth];
+            leftNodePositions = new int[treeDepth];
+            rightNodePositions = new int[treeDepth];
+            splitValuesStack = new byte[treeDepth][];
+            this.dataInput = new ByteArrayDataInput(packedIndex);
+            readNodeData(false);
+        }
+
+        public void reset()
+        {
+            nodeID = 1;
+            level = 0;
+            dataInput.setPosition(0);
+        }
+
+        public void pushLeft()
+        {
+            int nodePosition = leftNodePositions[level];
+            nodeID *= 2;
+            level++;
+            dataInput.setPosition(nodePosition);
+            readNodeData(true);
+        }
+
+        public void pushRight()
+        {
+            int nodePosition = rightNodePositions[level];
+            nodeID = nodeID * 2 + 1;
+            level++;
+            dataInput.setPosition(nodePosition);
+            readNodeData(false);
+        }
+
+        public void pop()
+        {
+            nodeID /= 2;
+            level--;
+        }
+
+        public boolean isLeafNode()
```

Comment:
nit: We could also call this `atLeafNode()` if you think it reads better...fine either way

Reply:
I agree, `atLeafNode` reads better because this is a traversal state so is telling us where we are.